boltons/ioutils.py

```diff
@@ -0,0 +1,322 @@
+
+    Example::
+
+        >>> from boltons import ioutils
+        >>> with ioutils.SpooledBytesIO() as f:
+        ...     f.write(b"Happy IO")
+        ...     _ = f.seek(0)
+        ...     isinstance(f.getvalue(), ioutils.binary_type)
+        True
+    """
+
+    def read(self, n=-1):
+        return self.buffer.read(n)
+
+    def write(self, s):
+        if not isinstance(s, binary_type):
+            raise TypeError("{0} expected, got {1}".format(
+                binary_type.__name__,
+                type(s).__name__
+            ))
+
+        self.buffer.write(s)
+        if self.tell() >= self._max_size:
+            self.rollover()
+
+    def readline(self, length=None):
+        return self.buffer.readline(length)
+
+    def readlines(self, sizehint=0):
+        return self.buffer.readlines(sizehint)
+
+    def rollover(self):
+        """Roll the StringIO over to a TempFile"""
+        if not self._rolled:
+            tmp = TemporaryFile()
+            pos = self._file.tell()
+            self.__buffer.seek(0)
+            tmp.write(self.__buffer.read())
+            tmp.seek(pos)
+            self.__buffer.close()
+            self.__buffer = tmp
+
+    @property
+    def _rolled(self):
+        return not isinstance(self.buffer, BytesIO)
+
+    @property
+    def buffer(self):
+        try:
+            return self.__buffer
+        except AttributeError:
+            self.__buffer = BytesIO()
+        return self.__buffer
+
+
+class SpooledStringIO(SpooledIOBase):
+    """
+    SpooledStringIO is a spooled file-like-object that only accepts unicode
+    values. On Python 2.x this means the 'unicode' type and on Python 3.x this
+    means the 'str' type. Values are accepted as unicode and then coerced into
+    utf-8 encoded bytes for storage. On retrieval, the values are returned as
+    unicode.
+
+    Example::
+
+        >>> from boltons import ioutils
+        >>> with ioutils.SpooledStringIO() as f:
+        ...     f.write(u"\u2014 Hey, an emdash!")
+        ...     _ = f.seek(0)
+        ...     isinstance(f.read(), ioutils.text_type)
+        True
+
+     """
+
+    def read(self, n=-1):
+        return self.buffer.read(n).decode('utf-8')
+
+    def write(self, s):
+        if not isinstance(s, text_type):
+            raise TypeError("{0} expected, got {1}".format(
+                text_type.__name__,
+                type(s).__name__
+            ))
+
+        self.buffer.write(s.encode('utf-8'))
+        if self.tell() >= self._max_size:
+            self.rollover()
+
+    def readline(self, length=None):
+        return self.buffer.readline(length).decode('utf-8')
+
+    def readlines(self, sizehint=0):
+        return [x.decode('utf-8') for x in self.buffer.readlines(sizehint)]
+
+    @property
+    def buffer(self):
+        try:
+            return self.__buffer
+        except AttributeError:
+            self.__buffer = EncodedFile(BytesIO(), data_encoding='utf-8')
```

Comment:
Seek is already implemented to operate on codepoints. I updated the pull request an implemented len and tell as well, but tell is pretty inefficient. I am looking at doing it in larger chunks when possible, something like 21333 which even if all 21333 characters are 3 btyes long, falls under 64000 byes, so it'll nicely fit into the L2 cache of most processors without any problems.

I'm thinking in terms of the tell implementation, a better option than what I did in my updated pull request would be to check the position (in bytes) then seek in chunks, recording each along the way, until the chunk read takes you past the original tell() point. At that point you could back up and read in smaller increments until you hit the byte position. That would be relatively efficient.